### PR TITLE
dev: cross compile binaries to speed up build time

### PR DESCRIPTION
**Pull request by ftupas**
Time spent on this PR: 0.2

Resolves: #425 

# Pull Request type

Please check the type of change your PR introduces:

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no API changes)
- [x] Build-related changes
- [ ] Documentation content changes
- [ ] Testing

# What is the new behavior?

Cross-compile the binaries and improve build time [57 m](https://github.com/alamatxyz/kakarot-rpc/actions/runs/5833654529/job/15821695313) &rarr; [12 m](https://github.com/alamatxyz/kakarot-rpc/actions/runs/5869304788/job/15914007439)

# Does this introduce a breaking change?

- [ ] Yes
- [x] No

# Added changes to [`CHANGELOG.md`](../CHANGELOG.md)

- [x] Yes
- [ ] No

### CODE REVIEW

File: .github/workflows/kakarot_rpc.yml [hunk 1]
```diff
@@ -11,7 +11,7 @@ env:
 
 jobs:
   build:
-    runs-on: ubuntu-latest-32-cores
+    runs-on: ubuntu-latest-16-cores
```
Comment (ftupas) [inline .github/workflows/kakarot_rpc.yml:14]: We can just use 16 cores finally with cross-compilation as it's fast enough https://github.com/alamatxyz/kakarot-rpc/actions/runs/5869304788/job/15914007439

File: Dockerfile [hunk 2]
```diff
@@ -1,15 +1,45 @@
...
+        LINKER=$3; \
+        echo "Building for $TARGETPLATFORM"; \
+         # Add the specified Rust target architecture
+        rustup target add $ARCH; \
+        # Update package lists and install the specified compiler
+        apt-get update && apt-get -y install $COMPILER; \
+        # Build the Rust application for the specified target
+        cargo build --all --release \
+          --target=$ARCH \
+          --config target.$ARCH.linker=\"$LINKER\"; \
```
Comment (ftupas) [inline Dockerfile:26]: Here we cross-compile to the target architecture, thanks @bajpai244 for the previous work with [katana](https://github.com/kkrt-labs/kakarot/pull/594/files)

File: crates/eth-rpc/Cargo.toml [hunk 3]
```diff
@@ -40,6 +40,9 @@ tracing = "0.1.34"
 tracing-subscriber = { version = "0.3.3", features = ["env-filter"] }
 dojo-test-utils = { workspace = true }
 
+# for cross-compiling
+openssl = { version = "0.10", features = ["vendored"] }
```
Comment (ftupas) [inline crates/eth-rpc/Cargo.toml:44]: This is needed so `openssl` is built from [source](https://docs.rs/openssl/latest/openssl/#vendored) for the right architecture

File: docker-compose.katana.yaml [hunk 4]
```diff
@@ -1,23 +1,23 @@
-version: "3.2"
+version: '3.2'
 
 services:
   starknet:
-    image: ghcr.io/dojoengine/dojo:latest
+    image: ghcr.io/dojoengine/dojo:nightly-5cbe549ec0d8c17fe1abbddea36bf55228043107
```
Comment (ftupas) [inline docker-compose.katana.yaml:5]: pin dojo as the latest changed from `--allow-zero-max-fee` to `--disable-fee`, will create an issue for this

**Review by Eikix — COMMENTED** [review]
lgtm but will let @bajpai244 approve:)

**Review by bajpai244 — APPROVED** [review]
Great work @ftupas 🎉 

Approved

**Pull request merged**